packages/cloud_firestore/cloud_firestore_platform_interface/lib/src/method_channel_cloud_firestore.dart

```diff
@@ -0,0 +1,439 @@
+      'DocumentReference#setData',
+      <String, dynamic>{
+        'app': app,
+        'path': path,
+        'data': data,
+        'options': options,
+      },
+    );
+  }
+
+  @override
+  Future<void> updateDocumentReferenceData(
+    String app, {
+    @required String path,
+    Map<String, dynamic> data = const <String, dynamic>{},
+  }) {
+    return channel.invokeMethod<void>(
+      'DocumentReference#updateData',
+      <String, dynamic>{
+        'app': app,
+        'path': path,
+        'data': data,
+      },
+    );
+  }
+
+  // TODO: Type this return
+  @override
+  Future<Map<String, dynamic>> getDocumentReference(
+    String app, {
+    @required String path,
+    @required String source,
+  }) {
+    return channel.invokeMapMethod<String, dynamic>(
+      'DocumentReference#get',
+      <String, dynamic>{
+        'app': app,
+        'path': path,
+        'source': source,
+      },
+    );
+  }
+
+  @override
+  Future<void> deleteDocumentReference(
+    String app, {
+    @required String path,
+  }) {
+    return channel.invokeMethod<void>(
+      'DocumentReference#delete',
+      <String, dynamic>{'app': app, 'path': path},
+    );
+  }
+
+  Future<int> _addDocumentReferenceSnapshotListener(
+    String app, {
+    @required String path,
+    bool includeMetadataChanges,
+  }) {
+    return channel.invokeMethod<int>(
+      'DocumentReference#addSnapshotListener',
+      <String, dynamic>{
+        'app': app,
+        'path': path,
+        'includeMetadataChanges': includeMetadataChanges,
+      },
+    );
+  }
+
+  static final Map<int, StreamController<int>> _documentObservers =
+      <int, StreamController<int>>{};
+
+  // This method is very similar to getQuerySnapshots. Extract common logic?
+  @override
+  Stream<dynamic> getDocumentReferenceSnapshots(
+    String app, {
+    @required String path,
+    bool includeMetadataChanges,
+  }) {
+    assert(includeMetadataChanges != null);
+    Future<int> _handle;
+    // It's fine to let the StreamController be garbage collected once all the
+    // subscribers have cancelled; this analyzer warning is safe to ignore.
+    StreamController<dynamic> controller; // ignore: close_sinks
+    controller = StreamController<dynamic>.broadcast(
+      onListen: () {
+        _handle = _addDocumentReferenceSnapshotListener(
+          app,
+          path: path,
+          includeMetadataChanges: includeMetadataChanges,
+        ).then<int>((dynamic result) => result);
+        _handle.then((int handle) {
+          _documentObservers[handle] = controller;
+        });
+      },
+      onCancel: () {
+        _handle.then((int handle) async {
+          await removeListener(handle);
+          _documentObservers.remove(handle);
+        });
+      },
+    );
+    return controller.stream;
+  }
+
+  void _handleDocumentSnapshot(MethodCall call) {
+    final int handle = call.arguments['handle'];
+    _queryObservers[handle].add(call.arguments);
```

Comment:
`_queryObservers` should be `_documentObservers`, right?

Reply:
Yep, good catch!